lib/internal/webstreams/adapters.js

```diff
@@ -0,0 +1,888 @@
+            // Value should always be undefined here.
+            readable.push(null);
+          } else {
+            readable.push(chunk.value);
+          }
+        },
+        (error) => readable.destroy(error));
+    },
+
+    destroy(error, callback) {
+      function done() {
+        try {
+          callback(error);
+        } catch (error) {
+          process.nextTick(() => { throw error; });
+        }
+      }
+
+      if (!closed) {
+        PromisePrototypeThen(
+          reader.cancel(error),
+          done,
+          done);
+        return;
+      }
+      done(error);
+    },
+  });
+
+  PromisePrototypeThen(
+    reader.closed,
+    () => {
+      closed = true;
+      if (!readable.readableEnded)
+        readable.push(null);
+    },
+    (error) => {
+      closed = true;
+      if (!readable.destroyed) {
+        readable.destroy(error);
+      }
+    });
+
+  return readable;
+}
+
+/**
+ * @typedef {import('./readablestream').ReadableWritablePair
+ * } ReadableWritablePair
+ * @typedef {import('../../stream').Duplex} Duplex
+ *
+ * @param {Duplex} duplex
+ * @returns {ReadableWritablePair}
+ */
+function newReadableWritablePairFromDuplex(duplex) {
+  if (typeof duplex?._writableState !== 'object' ||
+      typeof duplex?._readableState !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE('duplex', 'stream.Duplex', duplex);
+  }
+
+  const writable = newWritableStreamFromStreamWritable(duplex);
+  const readable = newReadableStreamFromStreamReadable(duplex);
+
+  return {
+    writable,
+    readable,
+  };
+}
+
+/**
+ * @param {ReadableWritablePair} pair
+ * @param {{
+ *   decodeStrings? : boolean,
+ *   encoding? : string,
+ *   highWaterMark? : number,
+ *   objectMode? : boolean,
+ *   signal? : AbortSignal,
+ * }} [options]
+ * @returns
+ */
+function newStreamDuplexFromReadableWritablePair(pair = {}, options = {}) {
+  validateObject(pair, 'pair');
+  const {
+    readable: readableStream,
+    writable: writableStream,
+  } = pair;
+
+  if (!isReadableStream(readableStream)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'pair.readable',
+      'ReadableStream',
+      readableStream);
+  }
+  if (!isWritableStream(writableStream)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'pair.writable',
+      'WritableStream',
+      writableStream);
+  }
+
+  validateObject(options, 'options');
+  const {
+    objectMode = false,
+    encoding,
+    decodeStrings = true,
+    highWaterMark,
+    signal,
+  } = options;
+
+  validateBoolean(objectMode, 'options.objectMode');
+  if (signal !== undefined)
+    validateAbortSignal(signal, 'options.signal');
+  if (encoding !== undefined && !Buffer.isEncoding(encoding))
+    throw new ERR_INVALID_ARG_VALUE(encoding, 'options.encoding');
+
+  const writer = writableStream.getWriter();
+  const reader = readableStream.getReader();
+  let writableClosed = false;
+  let readableClosed = false;
+
+  const duplex = new Duplex({
+    allowHalfOpen: false,
```

Comment:
Why not allowHalfOpen?

Reply:
Because in a paired ReadableStream/WritableStream the lifecycles are attached. There is no `allowHalfOpen` there.

Reply:
Isn't it 2 separate streams? e.g. it could be a body stream from fetch and a file write stream. Their lifecycles would not be attached?